### PR TITLE
Add ability to set preview time in editor

### LINKED ISSUE #14213: Editor has no way to set preview points
There should be a way to set preview points in the editor to reach feature parity with stable.
This is what it looks like in stable:

**osu!lazer version:** 2021.809.0-lazer

### PR DESCRIPTION
**Pull request by cdwcgt**
close #14213
I'm not sure if having it directly in editorBeatmap as a Bindable is the best way but it seems good

image

the green one is preview timing point(yellow already occupied by a non-inheritance point)

### CODE REVIEW

File: osu.Game/Screens/Edit/Editor.cs
Comment (bdach) [inline osu.Game/Screens/Edit/Editor.cs:330]: none of the other menu items are cased like this. i also don't know why this is talking about "position"

```suggestion
                                            new EditorMenuItem("Set current time as preview point", MenuItemType.Standard, SetCurrectTimeAsPreview)
```

File: osu.Game/Screens/Edit/EditorBeatmap.cs
Comment (bdach) [inline osu.Game/Screens/Edit/EditorBeatmap.cs:89]: should not be publically settable like this

File: osu.Game/Screens/Edit/Editor.cs
Comment (bdach) [inline osu.Game/Screens/Edit/Editor.cs:811]: typo in method name

File: osu.Game/Screens/Edit/EditorBeatmap.cs
Comment (bdach) [inline osu.Game/Screens/Edit/EditorBeatmap.cs:113]: isn't handling -1 correctly

https://github.com/ppy/osu/blob/60c8ef3fe5fe9adcb204814d975b7ed6c832fc8c/osu.Game/Beatmaps/BeatmapMetadata.cs#L48-L51

File: osu.Game/Screens/Edit/Components/Timelines/Summary/Parts/PreviewTimePart.cs
Comment (bdach) [inline osu.Game/Screens/Edit/Components/Timelines/Summary/Parts/PreviewTimePart.cs:28]: lime is already being used by difficulty points

https://github.com/ppy/osu/blob/60c8ef3fe5fe9adcb204814d975b7ed6c832fc8c/osu.Game/Beatmaps/ControlPoints/DifficultyControlPoint.cs#L31

**Review by bdach — CHANGES_REQUESTED** [review]

File: osu.Game/Screens/Edit/Editor.cs
Comment (Joehuu) [reply osu.Game/Screens/Edit/Editor.cs:330]: FWIW that is stable's wording/casing, but I agree with the suggestion above.

File: osu.Game/Screens/Edit/Components/Timelines/Summary/Parts/PreviewTimePart.cs
Comment (cdwcgt) [reply osu.Game/Screens/Edit/Components/Timelines/Summary/Parts/PreviewTimePart.cs:32]: I'm not sure which color to use, I need advice

File: osu.Game/Screens/Edit/Editor.cs
Comment (cdwcgt) [reply osu.Game/Screens/Edit/Editor.cs]: Maybe this needs to be discussed? Because I'm doing this to correspond with stable

File: osu.Game.Tests/Visual/Editing/TestScenePreviewTime.cs
Comment (bdach) [inline osu.Game.Tests/Visual/Editing/TestScenePreviewTime.cs:30]: why shouldn't it show? if `PreviewTime` is -1 then it's automatically 40% of the track length. surely you can show the line there

Comment (cdwcgt) [reply osu.Game.Tests/Visual/Editing/TestScenePreviewTime.cs:30]: This should be the expected situation.
The preview point should not be displayed on the map without preview point (also stable)

If it is really shown in the 40% position, it may be misleading (let mapper think its already set)
if we must do this, I think it should be a different color, but that would get weird

Comment (peppy) [reply osu.Game.Tests/Visual/Editing/TestScenePreviewTime.cs:30]: I think of these options, not showing at all is probably best.

File: osu.Game/Screens/Edit/Components/Timelines/Summary/Parts/PreviewTimePart.cs
Comment (LeNitrous) [reply osu.Game/Screens/Edit/Components/Timelines/Summary/Parts/PreviewTimePart.cs:32]: Stable uses yellow
image

Comment (bdach) [reply osu.Game/Screens/Edit/Components/Timelines/Summary/Parts/PreviewTimePart.cs:32]: not sure yellow is going to work either as lazer uses that for timing points

File: osu.Game/Screens/Edit/Components/Timelines/Summary/Parts/PreviewTimePart.cs
Comment (bdach) [inline osu.Game/Screens/Edit/Components/Timelines/Summary/Parts/PreviewTimePart.cs:16]: also just noticed that this is incorrect, you cannot pass bindables like this to component - make a local bindable in `PreviewTimeVisualisation` and bind to `beatmap.PreviewTime` instead

**Comment by bdach** [comment]
I've taken over this PR to hopefully bring it to a conclusion. Of note, while there are objectively correct fixes there (https://github.com/ppy/osu/pull/21650/commits/b689ad6d80e177734a470fc0d30e6497dc9af09b), I've also made subjective adjustments to the copy of the menu item to be more in line with everything else around it (can't care less about what stable's specific wording was), and also changed a few things in the visual appearance of the preview point marker in timeline.

Before:

image

After:

image

In particular:

- Marker uses slightly different hue of green to distinguish it from difficulty control points
- Marker no longer overlaps control point markers and is shown in the same manner as bookmarks are

Dunno if people will agree but these feel pretty no-brainer to me. Will need a second review.

**Review by bdach — DISMISSED** [review]

**Review by frenzibyte — APPROVED** [review]
This looks pretty fine to me. ~~I was hoping we can share the preview time at a `BeatmapSetInfo` level since we only allow one track per set, but that didn't turn out to be an easy change so I'm leaving it for later.~~

**Comment by bdach** [comment]
> I was hoping we can share the preview time at a `BeatmapSetInfo` level since we only allow one track per set

why would that even be in the scope of this change though. that is a wild swing of a thing to attempt without even trying to communicate prior intent? was this talked about anywhere?

**Comment by frenzibyte** [comment]
Considered it a small UX change that would definitely be nice to have rather than forcing the mapper to edit each difficulty to set the same preview point.

**Comment by bdach** [comment]
> Considered it a small UX change that would definitely be nice to have rather than forcing the mapper to edit each difficulty to set the same preview point

this is something to revisit with something like https://github.com/ppy/osu/issues/18903.

also i don't think stable enforces this and i don't know whether mappers don't want the freedom of setting preview point for whatever reason.

> (and it should be just using `BeatmapSetInfo.Metadata` in theory, but something with caching is breaking it from fully working).

that sounds incorrect because if this is to become a set-level operation then i'm pretty sure you need to re-encode all `.osu` files for all difficulties in the set to be fully correct operation. but again why are we even talking about this yet _without prior discussion_.

**Comment by frenzibyte** [comment]
Okay, let's just ignore it.

**Pull request merged**
